engine/storage/test/org/apache/cloudstack/storage/endpoint/DefaultEndPointSelectorTest.java

```diff
@@ -0,0 +1,167 @@
+import com.cloud.vm.SecondaryStorageVmVO;
+import com.cloud.vm.VirtualMachine;
+import com.cloud.vm.VirtualMachine.State;
+import com.cloud.vm.VirtualMachine.Type;
+import com.cloud.vm.dao.SecondaryStorageVmDao;
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest(ComponentContext.class)
+public class DefaultEndPointSelectorTest {
+    @Mock
+    HostDao _hostDao;
+
+    @Mock
+    SecondaryStorageVmDao _ssvmDao;
+
+    @InjectMocks
+    @Spy
+    DefaultEndPointSelector selector = new DefaultEndPointSelector();
+
+    //Special mocked endpoints that the tests can use to identify what was returned from the method.
+    private EndPoint selectMethodEndpoint = mock(EndPoint.class);
+    private DummyEndpoint dummyEndpoint = mock(DummyEndpoint.class);
+
+    /**
+     * The single common method that sets up the mocks to cover all possible scenarios we want to
+     * test. The VM type and volume state are passed in, while a potential list of existing SSVMs
+     * are also passed in. As its final action, the method calls {@link DefaultEndPointSelector#select(DataObject, StorageAction)}
+     * in order to test the method's operation and returns the endpoint found.
+     *
+     * @param vmType - State of the VM attached to the volume
+     * @param volumeState - State of the volume
+     * @param ssvmVOs - A list of SSVMs to return from the DAO
+     * @return The selected endpoint.
+     */
+    public EndPoint mockEndpointSelection(Type vmType, State volumeState, SecondaryStorageVmVO ... ssvmVOs) {
+        VolumeInfo volume = mock(VolumeInfo.class);
+        VirtualMachine vm = mock(VirtualMachine.class);
+        Long zoneId = 1l;
+
+        when(vm.getType()).thenReturn(vmType);
+        when(vm.getState()).thenReturn(volumeState);
+        when(volume.getAttachedVM()).thenReturn(vm);
+        when(volume.getDataCenterId()).thenReturn(zoneId);
+
+        List<SecondaryStorageVmVO> ssvmList = new ArrayList<>();
+        if (ssvmVOs != null) {
+            for (SecondaryStorageVmVO ssvm : ssvmVOs) {
+                ssvmList.add(ssvm);
+            }
+        }
+
+        when(_ssvmDao.listByZoneId(eq(SecondaryStorageVm.Role.templateProcessor), eq(zoneId)))
+            .thenReturn(ssvmList);
+
+        PowerMockito.mockStatic(ComponentContext.class);
+        when(ComponentContext.inject(DummyEndpoint.class)).thenReturn(dummyEndpoint);
+
+        doReturn(selectMethodEndpoint).when(selector).select(any(DataObject.class));
+
+        return selector.select(volume, StorageAction.DELETEVOLUME);
+    }
+
+    /**
+     * Dummy endpoint should be selected if an SSVM volume is expunging
+     * and no SSVMs are available.
+     */
+    @Test
+    public void testSelectionWithExpungingSsvm() {
+        EndPoint ep = mockEndpointSelection(Type.SecondaryStorageVm, State.Expunging);
+        Assert.assertNotNull(ep);
+        Assert.assertTrue(ep instanceof DummyEndpoint);
+        Assert.assertEquals(ep, dummyEndpoint);
```

Comment:
Lines 114 - 116 could be extracted to a method like the following:

```
private EndPoint executeTestSelectionWithExpunging(Type vmType, Object returnedEndPoint) {
    EndPoint ep = mockEndpointSelection(vmType, State.Expunging);
    Assert.assertNotNull(ep);
    Assert.assertEquals(ep, returnedEndPoint);
    return ep;
}
```

The endpoint returned by _executeTestSelectionWithExpunging(Type vmType, Object equalsTo)_ can be used in methods that perform additional tests (e.g. `Assert.assertTrue(ep instanceof DummyEndpoint);`).